File: bit manipulation/338_Counting Bits.py
from typing import List

#! time: O(nlog(n)) - space: O(n)
# class Solution:
#     def countBits(self, n: int) -> List[int]:
#         result = []
#         for i in range(0, n + 1):
#             count = 0
#             while i:
#                 if i & 1:
#                     count += 1
#                 i = i >> 1
#                 result.append(count)
#         return result

class Solution:
    def countBits(self, n: int) -> List[int]:
        

File: bit manipulation/338_Counting Bits.ts
// time: O(n) space: O(n)
function countBits(n: number): number[] {
    if (n === 0) {
        return [0];
    }
    const result: number[] = [];
    result.push(0, 1);
    let index = 1;
    let size = Math.floor((n + 1) / 2);
    for (let i = 1; i < size; ++i) {
        result.push(result[index]);
        result.push(result[index] + 1);
        index += 1;
    }
    if ((n + 1) % 2 !== 0) {
        result.push(result[index]);
    }
    return result;
};